### PR TITLE
OAuth Client - Add class `OAuth20ClientCredentialMixin` that allows to use OAuth2 client credential flow to clients

### PR DESCRIPTION
**Pull request by igorski-r7**
### Description
- Added the class `OAuth20ClientCredentialMixin` that allows to use OAuth2 client credential flow
- Updated the `requirements.txt` file

### Motivation and Context
- The change was required in the issue (https://issues.corp.rapid7.com/browse/PLGN-88)

### How Has This Been Tested?
- Tested it out locally with `make samples` on Microsoft plugins that uses the OAuth2 client credential flow

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [X] I have updated the documentation accordingly (if changes are required).
- [X] I have tested the changes locally.

### CODE REVIEW

File: insightconnect_plugin_runtime/clients/oauth.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    :param client_id: Client identifier required for OAuthSession
+    :type client_id: str
+
+    :param client_secret: Client Secret Key required to obtain the Authorization token
+    :type client_secret: str
+
+    :param token_url: Endpoint URL that allows you to obtain an authorization token
+    :type token_url: str
+    """
```
Comment (cmcnally-r7) [inline insightconnect_plugin_runtime/clients/oauth.py:21]: Should the description for these params be placed under the `__init__` class?

**Review by cmcnally-r7 — DISMISSED** [review]

File: README.md [hunk 2]
```diff
@@ -139,6 +139,7 @@ after cloning this repository.
 
 ## Changelog
 
+* 4.7.7 - Add `OAuth20ClientCredentialMixin` class to clients
```
Comment (dlaverty-r7) [inline README.md:142]: Should this update to 4.8.0 as its a new authentication method?

File: insightconnect_plugin_runtime/clients/oauth.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            **kwargs
+        )
+
+    @property
+    def auth_token(self) -> str:
+        """
+        The property allows you to refresh the authorization token
+        """
+
+        self.token = self.oauth.fetch_token(
```
Comment (dlaverty-r7) [inline insightconnect_plugin_runtime/clients/oauth.py:47]: Where/do we need to handle any errors/exceptions which arise from calling this function?

Comment (igorski-r7) [reply insightconnect_plugin_runtime/clients/oauth.py:47]: From the plugin perspective

**Review by llaszuk-r7 — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (igorski-r7) [reply README.md:142]: It's a new optional client - that adds optional functionality 😃

File: insightconnect_plugin_runtime/clients/oauth.py (hunk 3 above)
Comment (igorski-r7) [reply insightconnect_plugin_runtime/clients/oauth.py:47]: I will add the exception handler for `OAuth` class

File: tests/unit/test_oauth.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    "client_secret": "ExampleSecret",
+    "token_url": "https://example.com",
+}
+
+
+class TestOAuth2(unittest.TestCase):
+    @mock.patch("requests.Session.request")
+    @mock.patch(
+        "oauthlib.oauth2.BackendApplicationClient.parse_request_body_response",
+        return_value="ExmapleToken",
```
Comment (ablakley-r7) [inline tests/unit/test_oauth.py:24]:
```suggestion
        return_value="ExampleToken",
```

**Review by ablakley-r7 — COMMENTED** [review]
Looks great, one small comment on a value in testing.

**Review by cmcnally-r7 — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (dlaverty-r7) [reply README.md]: @igorski-r7 , I think that it should probably be treated as 4.8.0 as its a substantial update (even though optional) whereas I think leaving it at 4.7.7 implies its more of a defect fix?

File: tests/unit/test_oauth.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    def setUp(self, mock_request: Mock, mock_backend: Mock) -> None:
+        self.client = ExampleApiClass(**STUB_OAUTH_CLIENT_CREDENTIALS)
+
+    @mock.patch("requests_oauthlib.OAuth2Session.fetch_token")
+    def test_fetch_token(self, mock_fetch_token: Mock) -> None:
+        self.client.auth_token()
+        mock_fetch_token.assert_called_once()
+        mock_fetch_token.assert_called_with(**STUB_OAUTH_CLIENT_CREDENTIALS)
+
+    @mock.patch("requests_oauthlib.OAuth2Session.fetch_token")
```
Comment (dlaverty-r7) [inline tests/unit/test_oauth.py:35]: Can you add in a test for the failure scenario

**Review by dlaverty-r7 — APPROVED** [review]

**Review by cmcnally-r7 — APPROVED** [review]

**Pull request merged**